cpp/include/raft/spatial/knn/ball_cover.cuh

```diff
@@ -111,16 +126,71 @@ void rbc_all_knn_query(const raft::handle_t& handle,
   index.set_index_trained();
 }
 
+/**
+ * Performs a faster exact knn in metric spaces using the triangle
+ * inequality with a number of landmark points to reduce the
+ * number of distance computations from O(n^2) to O(sqrt(n)). This
+ * performs an all neighbors knn, which can reuse memory when
+ * the index and query are the same array. This function will
+ * build the index and assumes rbc_build_index() has not already
+ * been called.
+ * @tparam idx_t knn index type
+ * @tparam value_t knn distance type
+ * @tparam int_t type for integers, such as number of rows/cols
+ * @tparam matrix_idx_t matrix indexing type
+ * @param[in] handle raft handle for resource management
+ * @param[in] index ball cover index which has not yet been built
+ * @param[out] inds output knn indices
+ * @param[out] dists output knn distances
+ * @param[in] k number of nearest neighbors to find
+ * @param[in] perform_post_filtering if this is false, only the closest k landmarks
+ *                               are considered (which will return approximate
+ *                               results).
+ * @param[in] weight a weight for overlap between the closest landmark and
+ *               the radius of other landmarks when pruning distances.
+ *               Setting this value below 1 can effectively turn off
+ *               computing distances against many other balls, enabling
+ *               approximate nearest neighbors. Recall can be adjusted
+ *               based on how many relevant balls are ignored. Note that
+ *               many datasets can still have great recall even by only
+ *               looking in the closest landmark.
+ */
+template <typename idx_t = std::int64_t,
+          typename value_t,
+          typename int_t        = std::uint32_t,
+          typename matrix_idx_t = std::uint32_t>
+void rbc_all_knn_query(const raft::handle_t& handle,
+                       BallCoverIndex<idx_t, value_t, int_t, matrix_idx_t>& index,
+                       raft::device_matrix_view<idx_t, matrix_idx_t, row_major> inds,
+                       raft::device_matrix_view<value_t, matrix_idx_t, row_major> dists,
+                       int_t k                     = 5,
```

Comment:
I quite liked @cjnolet 's idea in my `raft::linalg` PR where we add another API for such booleans. Maybe that will be the cleanest here?

Reply:
@divyegala, the `linalg` case was a tad different because there were several options in addition to the booleans and the results of the booleans were each completely different algorithms. In this case, the boolean really is just a simple option. I think once we support an official "approximate" rbc, it would make sense to have 2 different functions.